README.rst
.. image:: https://raw.githubusercontent.com/molassesapp/molasses-go/main/logo.png
        :width: 100
        :height: 100
        :target: https://molasses.app

===============
molasses-python
===============

.. image:: https://img.shields.io/pypi/v/molasses.svg
        :target: https://pypi.python.org/pypi/molasses


A Python SDK for Molasses. It allows you to evaluate user's status for a feature. It also helps simplify logging events for A/B testing.

Molasses uses polling to check if you have updated features. Once initialized, it takes microseconds to evaluate if a user is active.


Install
-------

.. code:: python

  pip install molasses


Usage
-----

Initialization
~~~~~~~~~~~~~~

Start by initializing the client with an ``APIKey``. This begins the
polling for any feature updates. The updates happen every 15 seconds.

.. code:: python

   from molasses import MolassesClient

   client = MolassesClient("test_key")

If you decide not to track analytics events (experiment started,
experiment success) you can turn them off by setting the ``send_events``
field to ``False``

.. code:: python

   client = MolassesClient("test_key",  send_events=False)

Check if feature is active
~~~~~~~~~~~~~~~~~~~~~~~~~~

You can call ``is_active`` with the key name and optionally a user’s
information. The ``id`` field is used to determine whether a user is
part of a percentage of users. If you have other constraints based on
user params you can pass those in the ``params`` field.

.. code:: python

    client.is_active("FOO_TEST", {
      "id":"foo",
      "params":{
        "isBetaUser":"false",
        "isScaredUser":"false"
       }
    })

You can check if a feature is active for a user who is anonymous by just
calling ``is_active`` with the key. You won’t be able to do percentage
roll outs or track that user’s behavior.

.. code:: python

   client.is_active("TEST_FEATURE_FOR_USER")

Experiments
~~~~~~~~~~~

To track whether an experiment was successful you can call
``experiment_success``. experiment_success takes the feature’s name, any
additional parameters for the event and the user.

.. code:: python

   client.experiment_success("GOOGLE_SSO",{
           "version": "v2.3.0"
       },{
      "id":"foo",
      "params":{
        "isBetaUser":"false",
        "isScaredUser":"false"
       }
    })

Example
-------

.. code:: ruby

   from molasses import MolassesClient

   client = MolassesClient("test_key")

   if client.is_active('NEW_CHECKOUT'):
     print "we are a go"
   else:
     print "we are a no go"

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.1.0 (2020-09-24)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for molasses_python."""


tests/test_molasses_python.py
#!/usr/bin/env python

"""Tests for `molasses_python` package."""

import pytest

from molasses import MolassesClient

import responses

responseA = {
    "data": {
        "features": [
            {
                "active": True,
                "description": "foo",
                "key": "FOO_TEST",
                "segments": [
                    {
                        "constraint": "all",
                        "percentage": 100,
                        "segmentType": "alwaysControl",
                        "userConstraints": [
                            {
                                "userParam": "isScaredUser",
                                "operator": "in",
                                "values": "true,maybe",
                            },
                        ],
                    },
                    {
                        "constraint": "all",
                        "percentage": 100,
                        "segmentType": "alwaysExperiment",
                        "userConstraints": [
                            {
                                "userParam": "isBetaUser",
                                "operator": "equals",
                                "values": "true",
                            },
                        ],
                    },
                    {
                        "constraint": "all",
                        "percentage": 100,
                        "segmentType": "everyoneElse",
                        "userConstraints": [],
                    },
                ],
            },
        ],
    }, }
responseB = {
    "data": {
        "features": [
            {
                "id": "1",
                "active": True,
                "description": "foo",
                "key": "FOO_TEST",
                "segments": [
                    {
                        "constraint": "all",
                        "percentage": 100,
                        "segmentType": "alwaysControl",
                        "userConstraints": [
                            {
                                "userParam": "isScaredUser",
                                "operator": "nin",
                                "values": "false,maybe",
                            },
                        ],
                    },
                    {
                        "constraint": "all",
                        "percentage": 100,
                        "segmentType": "alwaysExperiment",
                        "userConstraints": [
                            {
                                "userParam": "isBetaUser",
                                "operator": "doesNotEqual",
                                "values": "false",
                            },
                        ],
                    },
                    {
                        "constraint": "all",
                        "percentage": 100,
                        "segmentType": "everyoneElse",
                        "userConstraints": [],
                    },
                ],
            },
        ],
    },
}

responseC = {
    "data": {
        "features": [
            {
                "id": "1",
                "active": True,
                "description": "foo",
                "key": "FOO_TEST",
                "segments": [
                    {
                        "percentage": 100,
                        "segmentType": "alwaysControl",
                        "constraint": "all",
                        "userConstraints": [
                            {
                                "userParam": "isScaredUser",
                                "operator": "contains",
                                "values": "scared",
                            },
                            {
                                "userParam": "isDefinitelyScaredUser",
                                "operator": "contains",
                                "values": "scared",
                            },
                            {
                                "userParam": "isMostDefinitelyScaredUser",
                                "operator": "contains",
                                "values": "scared",
                            },
                        ],
                    },
                    {
                        "percentage": 100,
                        "segmentType": "alwaysExperiment",
                        "constraint": "any",
                        "userConstraints": [
                            {
                                "userParam": "isBetaUser",
                                "operator": "doesNotContain",
                                "values": "fal",
                            },
                            {
                                "userParam": "isDefinitelyBetaUser",
                                "operator": "doesNotContain",
                                "values": "fal",
                            },
                        ],
                    },
                    {
                        "constraint": "all",
                        "percentage": 100,
                        "segmentType": "everyoneElse",
                        "userConstraints": [],
                    },
                ],
            },
        ],
    },
}

responseD = {
    "data": {
        "features": [
            {
                "id": "1",
                "active": True,
                "description": "foo",
                "key": "FOO_TEST",
                "segments": [],
            },
            {
                "id": "2",
                "active": False,
                "description": "foo",
                "key": "FOO_FALSE_TEST",
                "segments": [],
            },
            {
                "id": "3",
                "active": True,
                "description": "foo",
                "key": "FOO_50_PERCENT_TEST",
                "segments": [
                    {
                        "constraint": "all",
                        "segmentType": "everyoneElse",
                        "percentage": 50,
                        "userConstraints": [],
                    },
                ],
            },
            {
                "id": "4",
                "active": True,
                "description": "foo",
                "key": "FOO_0_PERCENT_TEST",
                "segments": [
                    {
                        "constraint": "all",
                        "segmentType": "everyoneElse",
                        "percentage": 0,
                        "userConstraints": [],
                    },
                ],
            },
            {
                "id": "5",
                "active": True,
                "description": "foo",
                "key": "FOO_ID_TEST",
                "segments": [
                    {
                        "constraint": "all",
                        "percentage": 100,
                        "segmentType": "alwaysControl",
                        "userConstraints": [
                            {
                                "userParam": "id",
                                "operator": "equals",
                                "values": "123",
                            },
                        ],
                    },
                    {
                        "constraint": "all",
                        "segmentType": "everyoneElse",
                        "percentage": 100,
                        "userConstraints": [],
                    },
                ],
            },
        ],
    },
}


@responses.activate
def test_basic():
    responses.add(responses.GET, 'https://us-central1-molasses-36bff.cloudfunctions.net/get-features',
                  json=responseA, status=200)

    molasses = MolassesClient("test_key",  send_events=False)
    assert molasses.is_active("FOO_TEST") is True
    assert molasses.is_active("FOO_TEST", {"foo": "foo"}) is True
    assert molasses.is_active("NOT_CHECKOUT") is False
    assert molasses.is_active("FOO_TEST", {"id": "foo", "params": {}}) is True
    assert molasses.is_active("FOO_TEST", {"id": "food", "params": {
                              "isScaredUser": "true"}}) is False
    assert molasses.is_active("FOO_TEST", {"id": "foodie", "params": {
                              "isBetaUser": "true"}}) is True


@responses.activate
def test_more_advanced():
    responses.add(responses.GET, 'https://us-central1-molasses-36bff.cloudfunctions.net/get-features',
                  json=responseB, status=200)

    molasses = MolassesClient("test_key",  send_events=False)
    assert molasses.is_active("FOO_TEST") is True
    assert molasses.is_active("FOO_TEST", {"foo": "foo"}) is True
    assert molasses.is_active("NOT_CHECKOUT") is False
    assert molasses.is_active("FOO_TEST", {"id": "foo", "params": {
                              "isBetaUser": "false", "isScaredUser": "false"}}) is True
    assert molasses.is_active("FOO_TEST", {"id": "food", "params": {
                              "isScaredUser": "true"}}) is False
    assert molasses.is_active("FOO_TEST", {"id": "foodie", "params": {
                              "isBetaUser": "true"}}) is True


@responses.activate
def test_even_more_advanced():
    responses.add(responses.GET, 'https://us-central1-molasses-36bff.cloudfunctions.net/get-features',
                  json=responseC, status=200)

    molasses = MolassesClient("test_key",  send_events=False)
    assert molasses.is_active("FOO_TEST", {"id": "foo", "params": {
        "isScaredUser": "scared",
        "isDefinitelyScaredUser": "scared",
        "isMostDefinitelyScaredUser": "scared",
    }}) is False
    assert molasses.is_active("FOO_TEST", {"id": "food", "params": {
                              "isDefinitelyBetaUser": "true", "isBetaUser": "true"}}) is True
    assert molasses.is_active("FOO_TEST", {"id": "foodie", "params": {
                              "isBetaUser": "true"}}) is True


@responses.activate
def test_percentage_tests():
    responses.add(responses.GET, 'https://us-central1-molasses-36bff.cloudfunctions.net/get-features',
                  json=responseD, status=200)

    molasses = MolassesClient("test_key", send_events=False)
    assert molasses.is_active("FOO_TEST") is True
    assert molasses.is_active("FOO_FALSE_TEST") is False
    assert molasses.is_active("FOO_50_PERCENT_TEST", {
                              "id": "140", "params": {}}) is False
    assert molasses.is_active("FOO_50_PERCENT_TEST", {
                              "id": "123", "params": {}}) is True
    assert molasses.is_active("FOO_0_PERCENT_TEST", {
        "id": "123", "params": {}}) is False
    assert molasses.is_active("FOO_ID_TEST", {
        "id": "123", "params": {}}) is False
    assert molasses.is_active("FOO_ID_TEST", {
        "id": "124", "params": {}}) is True
    assert molasses.is_active("FOO_ID_TEST", {
        "id": "122", "params": {}}) is True


@responses.activate
def test_experiments():
    responses.add(responses.GET, 'https://us-central1-molasses-36bff.cloudfunctions.net/get-features',
                  json=responseD, status=200)
    responses.add(responses.POST, 'https://us-central1-molasses-36bff.cloudfunctions.net/analytics',
                  json={}, status=200)

    molasses = MolassesClient("test_key", send_events=True)
    assert molasses.is_active("FOO_TEST") is True
    assert molasses.is_active("FOO_FALSE_TEST") is False
    assert molasses.is_active("FOO_50_PERCENT_TEST", {
                              "id": "140", "params": {}}) is False
    molasses.experiment_success("FOO_50_PERCENT_TEST", {}, {
        "id": "140", "params": {}})
    assert molasses.is_active("FOO_50_PERCENT_TEST", {
                              "id": "123", "params": {}}) is True
    molasses.experiment_success("FOO_50_PERCENT_TEST", {}, {
        "id": "123", "params": {}})


molasses/__init__.py
"""Top-level package for molasses."""

__author__ = """James Hrisho"""
__email__ = 'james.hrisho@gmail.com'
__version__ = '0.1.2'

import logging
import requests
import zlib
import json
from typing import Dict, Optional
from apscheduler.schedulers.background import BackgroundScheduler
from apscheduler.triggers.interval import IntervalTrigger
logger = logging.getLogger(__name__)

BASE_URL = 'https://us-central1-molasses-36bff.cloudfunctions.net'


class MolassesClient:
    """
    docstring
    """
    __cache = {}
    __initialized = False

    def __init__(self, api_key: str, send_events=True, base_url=BASE_URL):
        self.api_key = api_key
        self.send_events = send_events
        self.base_url = base_url
        self.__fetch_features()
        self.scheduler = BackgroundScheduler()
        self.features_job = self.scheduler.add_job(self.__fetch_features,
                                                   trigger=IntervalTrigger(seconds=int(15)))
        self.scheduler.start()

    def is_active(self, key: str, user: Optional[Dict] = None):
        if self.__initialized is not True:
            return False

        if key in self.__cache:
            feature = self.__cache[key]
            result = self.__is_active(feature, user)
            if user and "id" in user and self.send_events:
                self.__send_events({
                    "event": "experiment_started",
                    "tags": user["params"],
                    "userId": user["id"],
                    "featureId": feature["id"],
                    "featureName": key,
                    "testType": result if "experiment" else "control"
                })
            return result
        else:
            return False

    def experiment_success(self, key: str, additional_details: Dict, user: Optional[Dict] = None):
        if self.__initialized is not True or self.send_events is not True or user is None or "id" not in user:
            return False
        feature = self.__cache[key]
        result = self.__is_active(feature, user)
        self.__send_events({
            "event": "experiment_success",
            "tags": {**user["params"], **additional_details},
            "userId": user["id"],
            "featureId": feature["id"],
            "featureName": key,
            "testType": result if "experiment" else "control"
        })

    def stop(self):
        self.features_job.stop()
        self.scheduler.shutdown()

    def __is_active(self, feature, user=None):
        if feature["active"] is not True:
            return False
        if user is None or "id" not in user:
            return True
        segment_map = {}
        for feature_segment in feature["segments"]:
            segment_type = feature_segment["segmentType"]
            segment_map[segment_type] = feature_segment
        if "alwaysControl" in segment_map and self.__is_user_in_segment(user, segment_map["alwaysControl"]):
            return False
        if "alwaysExperiment" in segment_map and self.__is_user_in_segment(user, segment_map["alwaysExperiment"]):
            return True
        if "everyoneElse" in segment_map:
            return self.__get_user_percentage(user["id"], segment_map["everyoneElse"]["percentage"])
        return False

    def __get_user_percentage(self, id="", percentage=0):
        if percentage == 100:
            return True
        if percentage == 0:
            return False
        c = zlib.crc32(bytes(id, "utf-8")) & 0xffffffff
        v = abs(c % 100)
        return v < percentage

    def __is_user_in_segment(self, user: Optional[Dict], s: Dict):
        user_constraints = s["userConstraints"]
        constraints_length = len(user_constraints)
        constraints_to_be_met = 1 if s["constraint"] == "any" else constraints_length
        constraints_met = 0

        for i in range(constraints_length):
            constraint = user_constraints[i]
            param = constraint["userParam"]
            param_exists = param in user["params"]
            user_value = None
            if param_exists:
                user_value = user["params"][param]
            if param == "id":
                param_exists = True
                user_value = user["id"]
            if self.__meets_constraint(user_value, param_exists, constraint):
                constraints_met = constraints_met + 1
        return constraints_met >= constraints_to_be_met

    def __meets_constraint(self, user_value, param_exists, constraint):
        operator = constraint["operator"]
        if param_exists is False:
            return False
        if operator == "in":
            list_values = constraint["values"].split(",")
            return user_value in list_values
        elif operator == "nin":
            list_values = constraint["values"].split(",")
            return user_value not in list_values
        elif operator == "equals":
            return user_value == constraint["values"]
        elif operator == "doesNotEqual":
            return user_value != constraint["values"]
        elif operator == "contains":
            return user_value in constraint["values"]
        elif operator == "doesNotContain":
            return user_value not in constraint["values"]
        else:
            return False

    def __send_events(self, event_options: Dict):
        event_options["tags"] = json.dumps(event_options["tags"])
        requests.post(self.base_url + "/analytics", json=event_options, headers={
            "Authorization": "Bearer " + self.api_key
        })

    def __fetch_features(self):
        response = requests.get(self.base_url + "/get-features", params={}, headers={
            "Authorization": "Bearer " + self.api_key
        })
        if response.status_code == 200:
            data = response.json()
            if "data" in data:
                d = data.get("data")
                if "features" in d:
                    features = d.get("features")
                    for feature in features:
                        self.__cache[feature["key"]] = feature
                    self.__initialized = True
        else:
            logger.error("Molasses - %s %s",
                         response.status_code, response.text, exc_info=1)
